### PR TITLE
Fix #4163: always copy sliced data

### LINKED ISSUE #4163: [BLOCKING] /root/repo/xgboost/src/objective/regression_obj.cu:65: Check failed: preds.Size() == info.labels_.Size() (904 vs. 3616) labels are not correctly providedpreds.size=904, label.size=3616
1dac5e24108890c946c5a4673f0d1fa4d6125943 on fresh build off dmlc

[failxgbregression.zip](https://github.com/dmlc/xgboost/files/2881889/failxgbregression.zip)

@hcho3 seems to be **major** regression after recent fixes.  Seems all regression problems have the issue.

### PR DESCRIPTION
**Pull request by hcho3**
#4147 attempted to read sliced NumPy array directly into the C++ backend, but it relied on a NumPy internal variable (`data.base`) whose meaning we are not able to determine precisely. In particular, `len(data.base)` can sometimes mean the number of bytes in the NumPy array and other times mean the number of elements in the array.

Rather than dealing with undocumented NumPy internals, we'll take the approach of Microsoft/LightGBM#1210, i.e. always copy sliced data. This approach increases memory consumption slightly but the effect is not too bad. After all, the data matrix gets always copied when the input is a slice:
https://github.com/dmlc/xgboost/blob/c8c472f39aa08d5b8f180076a429f09317451cc7/python-package/xgboost/core.py#L487-L490

Closes #4163

### CODE REVIEW

**Comment by hcho3** [comment]
We'll need to investigate a better way to eliminate copies, per #4150.

File: python-package/xgboost/core.py [hunk 1]
```diff
@@ -637,15 +631,14 @@ def set_uint_info(self, field, data):
...
-        _check_call(_LIB.XGDMatrixSetUIntInfoStrided(
-            self.handle,
-            c_str(field),
-            c_data,
-            c_bst_ulong(stride),
-            c_bst_ulong(length)))
+        if getattr(data, 'base', None) is not None and \
+           data.base is not None and isinstance(data, np.ndarray) \
+           and isinstance(data.base, np.ndarray) and (not data.flags.c_contiguous):
+            raise ValueError('Sliced data is not allowed for uint property')
```
Comment (trivialfis) [inline python-package/xgboost/core.py]: Why not?

File: python-package/xgboost/core.py [hunk 2]
```diff
@@ -595,14 +584,15 @@ def set_float_info(self, field, data):
         data: numpy array
             The array of data to be set
         """
+        if getattr(data, 'base', None) is not None and \
+           data.base is not None and isinstance(data, np.ndarray) \
+           and isinstance(data.base, np.ndarray) and (not data.flags.c_contiguous):
+            raise ValueError('Should use set_float_info_npy2d() to use sliced data')
```
Comment (trivialfis) [inline python-package/xgboost/core.py]: Can we just redirect that to that call?

File: python-package/xgboost/core.py (hunk 1 above)
Comment (hcho3) [reply python-package/xgboost/core.py]: Right now, there is only one `uint` type field called `root_index`. Do we need slicing for this?

Comment (trivialfis) [reply python-package/xgboost/core.py]: We are here anyway, no need to leave something can be done in one line I think.  But go as you think best.

Comment (hcho3) [reply python-package/xgboost/core.py]: Addressed in the latest commit

**Comment by CodingCat** [comment]
this one is ready to go?
